src/lsm/forest.zig

```diff
@@ -162,22 +179,506 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+                    .compaction_blocks = self.compaction_blocks,
+                };
+            }
+
+            /// Our source and output blocks (excluding index blocks for now) are split two ways.
+            /// First, equally by pipeline stage, then by table a / table b:
+            /// -------------------------------------------------------------
+            /// | Pipeline 0                  | Pipeline 1                  |
+            /// |-----------------------------|-----------------------------|
+            /// | Table A     | Table B       | Table A     | Table B       |
+            /// -------------------------------------------------------------
+            fn divide_blocks(self: *CompactionPipeline) CompactionHelper.CompactionBlocks {
+                // Some blocks only need to be valid for a beat, after which they're used for
+                // the next compaction.
+                const minimum_block_count_beat: u64 = blk: {
+                    var minimum_block_count: u64 = 0;
+
+                    // We need a minimum of 2 source value blocks; one from each table.
+                    minimum_block_count += 2;
+
+                    // We need a minimum of 1 output value block.
+                    minimum_block_count += 1;
+
+                    // Because we're a 3 stage pipeline, with the middle stage (merge) having a
+                    // data dependency on both read and write value blocks, we need to split our
+                    // memory in the middle. This results in a doubling of what we have so far.
+                    minimum_block_count *= 2;
+
+                    break :blk minimum_block_count;
+                };
+
+                // Some blocks need to be reserved for the lifetime of the bar, and can't
+                // be shared between compactions, so these are all multipled by the number
+                // of trees:
+                const block_count_bar: u64 = blk: {
+                    var block_count: u64 = 0;
+
+                    // Source index blocks. For now, just require we have 2.
+                    block_count += 2;
+
+                    // Immutable value block. Used when level_a comes from the immutable table.
+                    block_count += 1;
+
+                    // Output index blocks. Minimum of 2 since one can be writing while the other
+                    // is being used.
+                    block_count += 2;
+
+                    block_count *= tree_infos.len;
+
+                    break :blk block_count;
+                };
+
+                assert(self.compaction_blocks.len >= block_count_bar + minimum_block_count_beat);
+
+                // Reserve the top of the block address space for the bar scoped blocks.
+                const blocks = self.compaction_blocks[0 .. self.compaction_blocks.len - block_count_bar];
+                assert(blocks.len >= minimum_block_count_beat);
+
+                for (blocks) |*block| block.next = null;
```

Comment:
Rather than nulling pointers here, _assert_ they are null, and null instead when the blocks are returned to the pool, to assert that the blocks have correct lifetimes.